contracts/adapters/OptimisticAuctionRebalanceExtensionV1.sol

```diff
@@ -0,0 +1,406 @@
+        bytes32 _rulesHash
+    )
+        external
+        onlyOperator
+    {
+        productSettings = ProductSettings({
+            optimisticParams: _optimisticParams,
+            rulesHash: _rulesHash
+        });
+        emit ProductSettingsUpdated(setToken, setToken.manager(), _optimisticParams, _rulesHash);
+    }
+
+     /**
+     * @param _quoteAsset                   ERC20 token used as the quote asset in auctions.
+     * @param _oldComponents                Addresses of existing components in the SetToken.
+     * @param _newComponents                Addresses of new components to be added.
+     * @param _newComponentsAuctionParams   AuctionExecutionParams for new components, indexed corresponding to _newComponents.
+     * @param _oldComponentsAuctionParams   AuctionExecutionParams for existing components, indexed corresponding to
+     *                                      the current component positions. Set to 0 for components being removed.
+     * @param _rebalanceDuration            Duration of the rebalance in seconds.
+     * @param _positionMultiplier           Position multiplier at the time target units were calculated.
+    */
+    function proposeRebalance(
+        IERC20 _quoteAsset,
+        address[] memory _oldComponents,
+        address[] memory _newComponents,
+        AuctionExecutionParams[] memory _newComponentsAuctionParams,
+        AuctionExecutionParams[] memory _oldComponentsAuctionParams,
+        uint256 _rebalanceDuration,
+        uint256 _positionMultiplier
+    )
+        external
+        onlyAllowedAssets(_newComponents)
+        onlyIfOpen()
+    {
+        bytes32 proposalHash = keccak256(abi.encode(
+            setToken,
+            _quoteAsset,
+            _oldComponents,
+            _newComponents,
+            _newComponentsAuctionParams,
+            _oldComponentsAuctionParams,
+            false, // We don't allow locking the set token in this version
+            _rebalanceDuration,
+            _positionMultiplier
+        ));
+        require(assertionIds[proposalHash] == bytes32(0), "Proposal already exists");
+        require(productSettings.rulesHash != bytes32(""), "Rules not set");
+        require(address(productSettings.optimisticParams.optimisticOracleV3) != address(0), "Oracle not set");
+
+
+        bytes memory claim = _constructClaim(proposalHash, productSettings.rulesHash);
+        uint256 totalBond = _pullBond(productSettings.optimisticParams);
+
+        bytes32 assertionId = productSettings.optimisticParams.optimisticOracleV3.assertTruth(
+            claim,
+            msg.sender,
+            address(this),
+            address(0),
+            productSettings.optimisticParams.liveness,
+            productSettings.optimisticParams.collateral,
+            totalBond,
+            productSettings.optimisticParams.identifier,
+            bytes32(0)
+        );
+
+        assertionIds[proposalHash] = assertionId;
```

Comment:
Let's double check unit tests for these lines

Reply:
Not sure what you'd like to test here exactly ?

Reply:
do we test that the claim is constructed correctly? or the bond is pulled?

Reply:
Will add test for pulling the bond, which is trivial. 
For testing the claim construction, right now it'd mean to reimplement the logic in the javascript tests and then compare the values, not sure how much value that'd bring though.

Reply:
@pblivin0x Or do you have another "source of truth" of what a "correctly constructed" claim should look like, to test against ?🤔

Reply:
See also:
https://github.com/IndexCoop/index-coop-smart-contracts/pull/156#discussion_r1441232920

Reply:
Added tests for both pulling the bond tand verifying the claim construction (by reproducing the same logic in JS) here:
https://github.com/IndexCoop/index-coop-smart-contracts/pull/157/commits/49bddfa9e343930b377d7d1cd5e8d1a7c9b94e9e